### PR TITLE
Add infill density benchmark

### PR DESCRIPTION
**Pull request by danieldouglas92**
Pull Request Checklist. Please read and check each box with an X. Delete any part not applicable. Ask on the [forum](https://community.geodynamics.org/c/aspect) if you need help with any step.

*Describe what you did in this PR and why you did it.*

### Before your first pull request:

* [x] I have read the guidelines in our [CONTRIBUTING.md](../blob/main/CONTRIBUTING.md) document.

### For all pull requests:

* [x] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [x] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

### CODE REVIEW

**Review by jdannberg — COMMENTED** [review]
These are the comments I made last night (so they are for an older version). I am submitting that review for now and then I'll have a look at the new version.

File: benchmarks/infill_density/infill_ascii_data.cc
Comment (jdannberg) [inline benchmarks/infill_density/infill_ascii_data.cc:227]: This looks very much like a description of a general ascii data boundary traction model. Can you add more details on what this specific mode does?

File: benchmarks/infill_density/infill_ascii_data.cc
Comment (jdannberg) [inline benchmarks/infill_density/infill_ascii_data.cc:146]: I am trying to understand this but I am not quite sure I do. Can you explain a little bit more the reasoning behind how the load from the ascii file is modified and what the assumptions are? Here is what I think you do: the idea is that you know the actual height of the volcano (but not the actual load), so you have to modify the load based on what the actual topography in the model is? If it's negative, you increase the load because you would have infill in the depression the load creates. And if it's positive, you decrease the load because some of it is already accounted for by the positive surface topography?

But I don't really understand the infill_height part. Why does this depend on the load, and not on the elevation?
Why does it use the `rock_density` rather than `sediment_density`? In the current file, it is zero, so for the benchmark this does not make a difference, but I imagine for an application model is does. So it would be good to explain this (and the bast place might be the class description, or the parameter descriptions).

**Review by jdannberg — COMMENTED** [review]
This is great, thank you for contributing this! If you have time, it would also be great if you could add a short description in the manual, but if not, that's fine as well. I mostly have a few suggestions for documentation, but otherwise this looks great!

File: benchmarks/infill_density/infill_ascii.prm
Comment (danieldouglas92) [reply benchmarks/infill_density/infill_ascii.prm:101]: Yes I did!

File: benchmarks/infill_density/infill_ascii_data.cc
Comment (danieldouglas92) [reply benchmarks/infill_density/infill_ascii_data.cc:146]: I updated the documentation but maybe it would be best if I also included another .png showing how this works? I'm finding it hard to explain in just text alone.

File: benchmarks/infill_density/infill_ascii.prm
Comment (naliboff) [inline benchmarks/infill_density/infill_ascii.prm:87]: Can you add a comment here (or in the section below) about why the left boundary is free-slip and the right is open? I assume this helps remove artifacts from return flow induced by the surface loading?

File: benchmarks/infill_density/infill_ascii.prm
Comment (naliboff) [inline benchmarks/infill_density/infill_ascii.prm:154]: Move this documentation down one line (suggestion below)

File: benchmarks/infill_density/infill_ascii.prm
Comment (naliboff) [inline benchmarks/infill_density/infill_ascii.prm:156]:
```suggestion
    # The first three values are assigned to the compositional fields representing viscoelastic 
    # stresses, but significantly these values are not used in the viscosity calculation (i.e., only
    # values representing lithologies are used). Respectively, the fourth and fifth values are 
    # assigned to the compositional fields representing the mantle and lithosphere.
    set Viscosities                 = 1.e21, 1.e21, 1.e21, 1.e21, 1.e26
```

File: benchmarks/infill_density/infill_ascii.prm
Comment (naliboff) [inline benchmarks/infill_density/infill_ascii.prm:106]: Out of curiosity, why does the GMT benchmark assume the rock and sediment density to be equal?

File: benchmarks/infill_density/infill_ascii_data.cc
Comment (naliboff) [inline benchmarks/infill_density/infill_ascii_data.cc:2]:
```suggestion
  Copyright (C) 2022 by the authors of the ASPECT code.
```

File: benchmarks/infill_density/infill_ascii_data.cc
Comment (naliboff) [reply benchmarks/infill_density/infill_ascii_data.cc:153]: The figure definitely helps. I don't have any suggestions to further improve the documentation above, and think it is in good shape.

File: benchmarks/infill_density/infill_ascii_data.cc
Comment (naliboff) [inline benchmarks/infill_density/infill_ascii_data.cc:186]: Could this and the patterns below just be `Patterns::Double (0.)`? In other words, the plugin is currently designed to only use one value for each and they are declared as doubles (rather than vectors of doubles).

**Review by naliboff — CHANGES_REQUESTED** [review]
@danieldouglas92 - Nice work, this is a really nice contribution! I only have minor questions and comments, and overall the PR is in great shape.

File: benchmarks/infill_density/infill_ascii_data.cc
Comment (danieldouglas92) [reply benchmarks/infill_density/infill_ascii_data.cc:186]: Ah yes good catch!

File: benchmarks/infill_density/infill_ascii.prm
Comment (danieldouglas92) [reply benchmarks/infill_density/infill_ascii.prm:83]: This actually was supposed to be set to normal but it does not seem to influence the results in any meaningful way

File: benchmarks/infill_density/infill_ascii.prm
Comment (danieldouglas92) [reply benchmarks/infill_density/infill_ascii.prm:106]: I honestly dont 100% understand why this is the case, but apparently the fourier solution is only exact with a constant density contrast between the load and the infill. If the infill varies then the fourier solution is still good but it is no longer 'exact' and now has to be approximated with a scaling prefactor.

File: benchmarks/infill_density/infill_ascii.prm
Comment (danieldouglas92) [reply benchmarks/infill_density/infill_ascii.prm:65]: I remember it increasing the accuracy yea, and my assumptions for why that was the case is what you added as a comment further down about better tracking for the lithosphere (since the analytic solution depends on lithosphere thickness).

**Review by jdannberg — COMMENTED** [review]
I think there were some comments from my first review you did not address (maybe because they were listed as outdated), I copied them here again. Otherwise this looks good!

I haven't looked at the manual text yet (since it does not build yet) and I can do this next after you rebase. 

The only other point I remembered: We usually have a test case for each benchmark. You don't need to copy the .prm file, but you can instead include your benchmark prm file in a test and then set the end time to zero (you can see an example of that in sol_kz_2_project_q1_only_visc.prm). The point is to make sure that even if we make future changes in ASPECT, the tester will still execute the files you added and we will notice if it breaks.

File: benchmarks/infill_density/data/box_2d_top.0.txt
Comment (danieldouglas92) [reply benchmarks/infill_density/data/box_2d_top.0.txt:1]: They are different! The load linearly increases with time, I found this helps stabilize the elastic response instead of dumping the entire load on all at once.

**Review by jdannberg — APPROVED** [review]
Great, looks good! Thank you for contributing this!
This is ready to merge once the tester is finished.

**Comment by jdannberg** [comment]
@naliboff You still have to approve before we can merge this.

File: benchmarks/infill_density/infill_ascii.prm
Comment (naliboff) [reply benchmarks/infill_density/infill_ascii.prm:66]: This parameter is actually independent from the DG composition formulation, but let's check it's affect later (probably has little to no influence).

**Review by naliboff — APPROVED** [review]
@danieldouglas92 - Nice work, ready to merge!

**Pull request merged**
